clients/pkg/promtail/targets/azureeventhub/parser.go

```diff
@@ -0,0 +1,198 @@
+package azureeventhub
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+type azureMonitorResourceLogs struct {
+	Records []json.RawMessage `json:"records"`
+}
+
+// validate check if message contains records
+func (l azureMonitorResourceLogs) validate() error {
+	if len(l.Records) == 0 {
+		return errors.New("records are empty")
+	}
+
+	return nil
+}
+
+// azureMonitorResourceLog used to unmarshal common schema for Azure resource logs
+// https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/resource-logs-schema
+type azureMonitorResourceLog struct {
+	Time          string `json:"time"`
+	Category      string `json:"category"`
+	ResourceID    string `json:"resourceId"`
+	OperationName string `json:"operationName"`
+}
+
+// validate check if fields marked as required by schema for Azure resource log are not empty
+func (l azureMonitorResourceLog) validate() error {
+	valid := len(l.Time) != 0 &&
+		len(l.Category) != 0 &&
+		len(l.ResourceID) != 0 &&
+		len(l.OperationName) != 0
+
+	if !valid {
+		return errors.New("required field or fields is empty")
+	}
+
+	return nil
+}
+
+type eventHubMessageParser struct {
+	disallowCustomMessages bool
+}
+
+func (e *eventHubMessageParser) Parse(message *sarama.ConsumerMessage, labelSet model.LabelSet, relabels []*relabel.Config, useIncomingTimestamp bool) ([]api.Entry, error) {
+	messageTime := time.Now()
+	if useIncomingTimestamp {
+		messageTime = message.Timestamp
+	}
+
+	data, err := e.tryUnmarshal(message.Value)
+	if err == nil {
+		err = data.validate()
+	}
+
+	if err != nil {
+		if e.disallowCustomMessages {
+			return []api.Entry{}, err
+		}
+
+		return []api.Entry{e.entryWithCustomPayload(message.Value, labelSet, messageTime)}, nil
```

Comment:
I'm not sure I follow exactly what we're trying to achieve here along the `disallowCustomMessages` condition.

If `tryUnmarshal` returns an error, then do we ever want to return something? In this case, what is `message.Value`, won't it be a nil byte slice? If so, is that a useful signal to have?

Reply:
Is the other case to populate the message in case it is _not_ a valid Azure Event Hub message but a random JSON object? Should we care about what's in there or allow anything to pass through?

Reply:
That is an excellent question 👍 

I think the title of this PR was misleading, and I already renamed it.

This target was created for Event Hubs. Therefore we can not guarantee or expect that there will be only structured logs in the messages. 

Customers can send messages(with custom logs or something else) to the Event Hub into any topic, even if cloud services use it for logs. Therefore, there might be a mix of structured and unstructured data.

Allowing custom message format gives customers more flexibility. Having a `disallowCustomMessages` flag will enable them to filter unstructured data if they would like.